app/services/moderator/merge_user.rb

```diff
@@ -55,45 +58,77 @@ def update_social
       end
     end
 
+    def merge_sponsorships
+      @delete_user.sponsorships.update_all(user_id: @keep_user.id)
+    end
+
     def merge_profile
-      if @delete_user.github_repos.any?
-        @delete_user.github_repos.update_all(user_id: @keep_user.id)
-        @keep_user.touch(:github_repos_updated_at)
-      end
-      if @delete_user.badge_achievements.any?
-        @delete_user.badge_achievements.update_all(user_id: @keep_user.id)
-        BadgeAchievement.counter_culture_fix_counts(where: { users: { id: @keep_user.id } })
-      end
+      @delete_user.github_repos.update_all(user_id: @keep_user.id)
+      @keep_user.touch(:github_repos_updated_at)
 
       @keep_user.update_columns(created_at: @delete_user.created_at) if @delete_user.created_at < @keep_user.created_at
     end
 
-    def merge_chat_mentions
-      any_memberships = @delete_user.chat_channel_memberships.any?
-      @delete_user.chat_channel_memberships.update_all(user_id: @keep_user.id) if any_memberships
-      @delete_user.mentions.update_all(user_id: @keep_user.id) if @delete_user.mentions.any?
+    def merge_badge_achievements
+      duplicate_badges = @delete_user.badge_achievements.pluck(:badge_id).intersection(
+        @keep_user.badge_achievements.pluck(:badge_id),
+      )
+      @delete_user.badge_achievements.where(badge_id: duplicate_badges).delete_all
+      @delete_user.badge_achievements.update_all(user_id: @keep_user.id)
+      BadgeAchievement.counter_culture_fix_counts(where: { users: { id: @keep_user.id } })
     end
 
-    def merge_follows
-      @delete_user.follows&.update_all(follower_id: @keep_user.id) if @delete_user.follows.any?
-      @delete_user_followers = Follow.followable_user(@delete_user.id)
-      @delete_user_followers.update_all(followable_id: @keep_user.id) if @delete_user_followers.any?
+    def merge_chat_channels
+      duplicate_moderator_memberships = @delete_user.chat_channel_memberships.where(
+        status: "active", role: "mod",
+      ).pluck(:chat_channel_id).intersection(
+        @keep_user.chat_channel_memberships.where(
+          status: "active", role: "mod",
+        ).pluck(:chat_channel_id),
+      )
+      @delete_user.chat_channel_memberships.where(chat_channel_id: duplicate_moderator_memberships).delete_all
```

Comment:
side note - because ChatChannelMembership includes Searchable (the record is elasticsearch indexed), it looks like there was a callback on destroy to remove from elasticsearch, is it safe to call delete_all here (rather than destroy all, for example)? Destroy all will definitely instantiate one by one, then call destroy and all associated callbacks. If that's a performance hit we can't take, is there a good way to bulk remove from ES (it doesn't look like can we pass an array of ids to `Search::RemoveFromIndexWorker` - but writing a bulk option (where id can be an array or a single id) doesn't seem too difficult).

Reply:
I'm hypothesizing that it doesn't matter, based on these assumptions:

- Our searches in Elasticsearch return a list of ids that we then query by
- If any of these ids no longer exists, it simply won't be returned in the result set from Postgres
- Leaving these records in Elasticsearch bloats the index, but we're in the process of removing our reliance on it anyway, so what is or is not in the index won't matter, so we only have to deal with the index bloat temporarily

These are all assumptions, so if any are wrong, my hypothesis that it doesn't matter may also be wrong. 😄